pkg/cloudprovider/providers/gce/gce_instances.go

```diff
@@ -250,50 +253,56 @@ func (gce *GCECloud) AddSSHKeyToAllInstances(user string, keyData []byte) error
 	})
 }
 
-// GetAllZones returns all the zones in which nodes are running
-func (gce *GCECloud) GetAllZones() (sets.String, error) {
-	// Fast-path for non-multizone
-	if len(gce.managedZones) == 1 {
-		return sets.NewString(gce.managedZones...), nil
-	}
-
-	// TODO: Caching, but this is currently only called when we are creating a volume,
-	// which is a relatively infrequent operation, and this is only # zones API calls
+// GetAllCurrentZones returns all the zones in which nodes are currently running
+func (gce *GCECloud) GetAllCurrentZones() (sets.String, error) {
 	zones := sets.NewString()
+	client := gce.client
+	if client == nil {
+		return nil, fmt.Errorf("Client not set on GCECloud object")
+	}
+	nodeList, err := client.Core().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		log.Fatalf("Failed to list nodes: %v", err)
+		return nil, err
+	}
 
-	// TODO: Parallelize, although O(zones) so not too bad (N <= 3 typically)
-	for _, zone := range gce.managedZones {
-		mc := newInstancesMetricContext("list", zone)
-		// We only retrieve one page in each zone - we only care about existence
-		listCall := gce.service.Instances.List(gce.projectID, zone)
-
-		// No filter: We assume that a zone is either used or unused
-		// We could only consider running nodes (like we do in List above),
-		// but probably if instances are starting we still want to consider them.
-		// I think we should wait until we have a reason to make the
-		// call one way or the other; we generally can't guarantee correct
-		// volume spreading if the set of zones is changing
-		// (and volume spreading is currently only a heuristic).
-		// Long term we want to replace GetAllZones (which primarily supports volume
-		// spreading) with a scheduler policy that is able to see the global state of
-		// volumes and the health of zones.
-
-		// Just a minimal set of fields - we only care about existence
-		listCall = listCall.Fields("items(name)")
-		res, err := listCall.Do()
-		if err != nil {
-			return nil, mc.Observe(err)
-		}
-		mc.Observe(nil)
-
-		if len(res.Items) != 0 {
+	for _, node := range nodeList.Items {
+		labels := node.ObjectMeta.Labels
+		zone, ok := labels[kubeletapis.LabelZoneFailureDomain]
+		if ok {
 			zones.Insert(zone)
+		} else {
+			return nil, fmt.Errorf("Could not find zone for node %v", node.ObjectMeta.Name)
 		}
 	}
-
 	return zones, nil
 }
 
+// GetAllManagedZones gets all the zones managed by k8s
+func (gce *GCECloud) GetAllManagedZones() (sets.String, error) {
+	if gce.managedZones != nil {
```

Comment:
Yes, it set from the config and doesn't ever change after. I feel like the zones returned by this method could be different from GetAllCurrentZones

Reply:
I'm not sure how useful "zones when initialized" is considering you can add/remove zones during runtime.

Reply:
Yes - that was my point when I suggested removing it. It seems @davidz627 agreed for removing this function completely in another comment I added.

Reply:
Something that confuses me.  There are a few places that iterate through managedZones, but if managedZones never changed after initialization, how does it get new zones that are added later?

Reply:
If you look into the code carefully managedZones can only be set to:
1. single zone (in case of zonal clusters)
2. all zones from the region (in case of multizonal ones).

So there is no need to change that.

Reply:
Ah got it thanks!

Reply:
I am a little confused as well. In a multizonal cluster then we have this:

https://github.com/kubernetes/kubernetes/blob/6b6b1e5779449a63079c9ff400da752975fe8a06/pkg/cloudprovider/providers/gce/gce_disks.go#L279

Which would technically allow creation of PDs in all zones in the region. Is this expected behavior? It is my understanding that this PR was meant to solve this problem (which it does by never supplying this function with a zone set that contains no nodes), however the check I linked above does not seem to be consistent with our desired behavior.

Reply:
It seems like all other uses of managedZones makes sense to me. Just this one seems to be related to the issue this PR is trying to solve. Can anyone confirm?

Reply:
It's checking that the zone passed in is in the managedZones list, but maybe it makes more sense to compare against GetAllCurrentZones.  The actual selection of the zone is done by ChooseZoneForVolumes, which calls GetAllCurrentZones already.  But it's also possible for the user to explicitly set an invalid zone as well.

Reply:
@saad-ali we were wondering if we should restrict both CreateDisk and CreateRegionalDisk to only being able to create disks in zones with nodes in them (a la GetAllCurrentZones). Or if they should be left in the current state (to allow users to create disks in any zone in the region).

In its current state we believe that a user trying to create a Disk in a zone without a node would be most likely a mistake (since the PD would be unusable) and should trigger an error. However, it could be possible that advanced users may want to be able to create PDs in zones without nodes for regionaldisks for manual failover.

What do you think, Saad? It seems to be a tradeoff between failing fast for standard use-cases, or allowing users to do hypothetical funky things manually that we may not want.

Reply:
FWIW @msau42 and I think surfacing an error for provisioning a PD in a zone without nodes is the way to go here. But I am interested to see if this may have some unintended side effects or if we missed some other use cases.

Reply:
> @saad-ali we were wondering if we should restrict both CreateDisk and CreateRegionalDisk to only being able to create disks in zones with nodes in them (a la GetAllCurrentZones). Or if they should be left in the current state (to allow users to create disks in any zone in the region).

In its current state we believe that a user trying to create a Disk in a zone without a node would be most likely a mistake (since the PD would be unusable) and should trigger an error. However, it could be possible that advanced users may want to be able to create PDs in zones without nodes for regionaldisks for manual failover.

What do you think, Saad? It seems to be a tradeoff between failing fast for standard use-cases, or allowing users to do hypothetical funky things manually that we may not want.
> 
> FWIW @msau42 and I think surfacing an error for provisioning a PD in a zone without nodes is the way to go here. But I am interested to see if this may have some unintended side effects or if we missed some other use cases.

I agree, provisioning a volume in a zone that does not have any active nodes should result in an error.